packages/z2s/src/sql.ts

```diff
@@ -24,32 +24,63 @@ export function formatSqlite(sql: SQLQuery) {
 }
 
 const sqlConvert = Symbol('fromJson');
-type SqlConvertArg = {
-  [sqlConvert]: true;
-  type: ValueType;
-  value: unknown;
-  plural: boolean;
-};
+export type LiteralType = 'boolean' | 'number' | 'string' | 'null';
+export type PluralLiteralType = Exclude<LiteralType, 'null'>;
+
+type SqlConvertArg =
+  | {
+      [sqlConvert]: 'column';
+      type: string;
+      value: unknown;
+      plural: boolean;
+    }
+  | {
+      [sqlConvert]: 'literal';
+      type: LiteralType;
+      value: LiteralValue;
+      plural: boolean;
+    };
+
 function isSqlConvert(value: unknown): value is SqlConvertArg {
   return value !== null && typeof value === 'object' && sqlConvert in value;
 }
 
-export function sqlConvertArg<T extends ValueType, P extends boolean = false>(
-  type: T,
-  value: true extends P ? TypeNameToTypeMap[T][] : TypeNameToTypeMap[T],
-  // plural is an explicit argument so we do not get confused by a singular JSON array vs
-  // an array of JSON.
-  plural?: P,
+export function sqlConvertSingularLiteralArg(
+  value: string | boolean | number | null,
 ): SQLQuery {
-  return sql.value({[sqlConvert]: true, type, value, plural});
+  const arg: SqlConvertArg = {
+    [sqlConvert]: 'literal',
+    type: value === null ? 'null' : (typeof value as LiteralType),
+    value,
+    plural: false,
+  };
+  return sql.value(arg);
 }
 
-export function sqlConvertArgUnsafe(
-  type: ValueType,
+export function sqlConvertPluralLiteralArg(
+  type: PluralLiteralType,
+  value: PluralLiteralType[],
+): SQLQuery {
+  const arg: SqlConvertArg = {
+    [sqlConvert]: 'literal',
+    type,
+    value,
+    plural: false,
```

Comment:
Maybe this'll make sense to me as I review but `plural: false` inside of `sqlConvertPluralLiteralArg` is surprising.